src/rmp/src/slack_tuning_strategy.cpp

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2025-2025, The OpenROAD Authors
+
+#include "slack_tuning_strategy.h"
+
+#include <cstddef>
+#include <cstdlib>
+#include <random>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "cut/abc_library_factory.h"
+#include "db_sta/dbNetwork.hh"
+#include "db_sta/dbSta.hh"
+#include "gia.h"
+#include "odb/db.h"
+#include "sta/Delay.hh"
+#include "sta/MinMax.hh"
+#include "utils.h"
+#include "utl/Logger.h"
+#include "utl/unique_name.h"
+
+namespace rmp {
+
+// The magic numbers are defaults from abc/src/base/abci/abc.c
+constexpr size_t SEARCH_RESIZE_ITERS = 100;
+constexpr size_t FINAL_RESIZE_ITERS = 1000;
+
+using utl::RMP;
+
+std::string SolutionSlack::toString() const
+{
+  if (solution_.empty()) {
+    return "[]";
+  }
+
+  std::ostringstream resStream;
+  resStream << '[';
+  if (!solution_.empty()) {
+    resStream << std::to_string(solution_.front().id);
+    for (int i = 1; i < solution_.size(); i++) {
+      resStream << ", " << std::to_string(solution_[i].id);
+    }
+  }
+  resStream << "], worst slack: ";
+
+  if (worst_slack_) {
+    resStream << *worst_slack_;
+  } else {
+    resStream << "not computed";
+  }
+  return resStream.str();
+}
+
+bool SolutionSlack::operator<(const SolutionSlack& other) const
+{
+  // Highest slack first
+  return worst_slack_ > other.worst_slack_;
+}
+
+SolutionSlack::Type SolutionSlack::RandomNeighbor(
+    const SolutionSlack::Type& all_ops,
+    utl::Logger* logger,
+    std::mt19937& random) const
+{
+  SolutionSlack::Type sol = solution_;
+  enum Move
+  {
+    ADD,
+    REMOVE,
+    SWAP,
+    COUNT
+  };
+  Move move = ADD;
+  if (sol.size() > 1) {
+    move = Move(random() % (COUNT));
+  }
+  switch (move) {
+    case ADD: {
+      debugPrint(logger, RMP, "annealing", 2, "Adding a new GIA operation");
+      size_t i = random() % (sol.size() + 1);
+      size_t j = random() % all_ops.size();
+      sol.insert(sol.begin() + i, all_ops[j]);
+    } break;
+    case REMOVE: {
+      debugPrint(logger, RMP, "annealing", 2, "Removing a GIA operation");
+      size_t i = random() % sol.size();
+      sol.erase(sol.begin() + i);
+    } break;
+    case SWAP: {
+      debugPrint(
+          logger, RMP, "annealing", 2, "Swapping adjacent GIA operations");
+      size_t i = random() % (sol.size() - 1);
+      std::swap(sol[i], sol[i + 1]);
+    } break;
+    case COUNT:
+      // unreachable
```

Comment:
I would add a `__builtin_unreachable` here so you can optimize out this case.   You get slightly nicer asm https://godbolt.org/z/1ehnhqrcx, given that it actually is unreachable.   Clang and gcc both have __builtin_unreachable so I think it's very unlikely to encounter a situation where it doesn't work.

Reply:
I used `ABSL_UNREACHABLE` from abseil. The preferable approach, `std::unreachable`, will be available since c++23.

Macro is defined here: https://github.com/abseil/abseil-cpp/blob/88c48235c8add3442800a19d5a7ee639dc242afa/absl/base/optimization.h#L207